### ISSUE TITLE
'pnpm run' kills child processes prematurely upon SIGINT

### ISSUE THREAD

**Issue by benweint** [pnpm/pnpm#7374]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Link to the code that reproduces this issue or a replay of the bug

https://gist.github.com/benweint/fb6ee6813347e1211d6d809ba33efedc

### Reproduction steps

Create a `package.json` with a `scripts` entry which runs a script with a signal handler, where the signal handler takes some non-trivial amount of time to run (say, 1s). Run the script via `pnpm run`. Observe that the child process spawned for the script exits before the signal handler has a chance to complete.

Example:

```
❯ cat package.json 
{
  "name": "pnpm-signals",
  "private": true,
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "dev": "node signals.js"
  },
  "author": "",
  "license": "ISC"
}

❯ cat signals.js 
process.once('SIGINT', () => {
  console.log('caught SIGINT, will exit in 1s');

  setTimeout(() => {
    console.log('exiting after 1s')
    process.exit(0);
  }, 1_000);
});

setInterval(() => {
  console.log('waiting ...');
}, 1_000);

❯ pnpm run dev

> pnpm-signals@1.0.0 dev /Users/ben/Desktop/pnpm-signals
> node signals.js

waiting ...
^Ccaught SIGINT, will exit in 1s
```

Note that the 'exiting after 1s' message was never seen. Compare to the behavior when running the script directly:

```
❯ node signals.js 
waiting ...
^Ccaught SIGINT, will exit in 1s
waiting ...
exiting after 1s
```

... or via `npm`:

```
❯ npm run dev

> pnpm-signals@1.0.0 dev
> node signals.js

waiting ...
^Ccaught SIGINT, will exit in 1s
waiting ...
exiting after 1s
```

### Describe the Bug

Note that this is adapted from https://github.com/pnpm/pnpm/issues/2653#issuecomment-1476686711, since that issue is marked as closed.

When the `package.json` file contains a `scripts` entry that has a signal handler which takes a non-trivial amount of time to execute, `pnpm` appears to terminate the child process before the signal handler has actually had time to complete. This is in contrast to the behavior of invoking the child process directly from a shell, and from the behavior of `npm run` (in both cases, the signal handler runs to completion).

This matters in cases where the child process needs to do some cleanup on shutdown in order to no leave behind state (for example, signaling children of its own who have been started in their own process groups).

### Expected Behavior

I would expect `pnpm run` to wait for child processes to actually exit before killing them.

### Which Node.js version are you using?

v20.4.0

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by gajus** [comment]
Thank you for raising this

**Comment by benweint** [comment]
Here's what I've learned from investigating this so far. The relevant signal handling code appears to live [here](https://github.com/pnpm/npm-lifecycle/blob/main/index.js#L329-L335) in pnpm/npm-lifecycle.

If you send *only* the `pnpm` command itself a `SIGINT` by using, say `kill -INT <pid-of-pnpm>` from another terminal window, things work as desired. However, when you press ctrl-C in a terminal, the shell actually sends a `SIGINT` not just to pnpm, but rather to all process in the foreground process group. `pnpm` then forwards *another* `SIGINT` in this situation to the child via [this line](https://github.com/pnpm/npm-lifecycle/blob/main/index.js#L330).

You can emulate the ctrl-C handling by sending `SIGINT` to the entire process group which `pnpm` creates, with `kill -INT -<pid-of-pnpm>`.

It seems npm has since moved their script running code into https://github.com/npm/run-script. https://github.com/npm/run-script/pull/142 changed their code to no longer forward signals to child processes at all, which spawned a handful of other bugs (e.g. https://github.com/npm/cli/issues/6684) for cases wherein folks are using `npm` outside the context of an interactive terminal session and expecting a `SIGINT` or `SIGTERM` sent to the `npm` process to be forwarded to children.

**Comment by gajus** [comment]
For context, the way I suggest pnpm handles this is...

1. first time user issues `ctrl-C`, pnpm should fire `SIGINT` to the child processes and wait for them to gracefully terminate
2. second time user issues `ctrl-C`, pnpm sends `SIGKILL` to the child processes and immediately exits

**Comment by benweint** [comment]
@gajus I believe this is actually the current behavior of `pnpm`!

The trouble here is that the `SIGINT` is actually delivered *twice*, and the example script in the repro case configures a signal handler with `process.once` rather than `process.on`. Breaking that down:

In POSIX systems, when you press ctrl-C in a terminal, the terminal delivers `SIGINT` not just to the process you directly started, but to the entire [foreground process group](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap11.html#tag_11_01_02). This behavior is specified in POSIX [§11.1.9](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap11.html#tag_11_01_09). In the case of `pnpm run`, since it does not use the `detached` [option](https://nodejs.org/api/child_process.html#optionsdetached) when spawning child processes to run scripts, both `pnpm` itself and any child processes it spawns to run scripts will be part of the same (foreground) process group when running in an interactive terminal.

So, when running `pnpm run` in an interactive terminal session, and then pressing ctrl-C, the sequence of events is as follows:

1. The terminal notices the ctrl-C and sends `SIGINT` to the entire foreground process group.
2a. In the child process, the callback provided to `process.once('SIGINT', ...)` fires, and then the [default `SIGINT` handler](https://nodejs.org/api/process.html#signal-events) (which simply exits the process) will be restored. [This step happens 'in parallel' with 2b.]
2b. In the `pnpm` process, [this SIGINT signal handler](https://github.com/pnpm/npm-lifecycle/blob/main/index.js#L329-L335) runs, which forwards a *second* `SIGINT` signal to the child, waits for the child to exit, and installs another `SIGINT` handler in the `pnpm` process which will send `SIGKILL` to the child if `pnpm` receives another `SIGINT`.
3. The second `SIGINT` is delivered to the child process, which causes the default `SIGINT` handler to run, and it makes the process exit.

`npm` opted to solve this problem of double-delivery by just *not* forwarding signals to children, but this seems to me (and others in https://github.com/npm/cli/issues/6684) a step too far - it means that signals sent to the parent `npm` process in non-internative-terminal contexts don't get forwarded to children, which can lead to all kinds of troubles for graceful shutdown in those non-interactive contexts.

Several thoughts on possible solutions:

1. If changing the child process is possible, just modify it to use `process.on` instead of `process.once` for its `SIGINT` handler.
2. It *might* be reasonable to have `pnpm` decide whether to do signal forwarding to child processes by attempting to determine whether it's in an interactive terminal context (e.g. by checking whether stdout is a TTY), but I'm not sure.
3. It *might* also be reasonable to expose a flag or environment variable to control the signal forwarding behavior.

**Comment by zaripych** [comment]
I'm experiencing similar issue. Though I'm using `process.on` instead.

Here is the repro link https://stackblitz.com/edit/node-qxscsn?file=index.ts

**Comment by teebot** [comment]
> Thank you for raising this. Despite #2653 being closed, this remains a major inconvenience.

yup. Caused a major issue here where a transaction was interrupted abruptly causing state inconsistencies. 
We ended up starting only with node instead of a pnpm script

**Comment by lucasbemol** [comment]
Hey Guys, any solution?

**Comment by mo-fouad** [comment]
@lucasbemol  you can add this to avoid any inconvenience. 

<code>
  // Handle SIGINT to gracefully shutdown
  process.on('SIGINT', () => {
    console.log('Gracefully shutting down from SIGINT (Ctrl+C)')
    server.close(() => {
      console.log('Server shut down');
      process.exit(0);
    });
  });
</code>

**Comment by skyrpex** [comment]
> I'm experiencing similar issue. Though I'm using `process.on` instead.
> 
> Here is the repro link https://stackblitz.com/edit/node-qxscsn?file=index.ts

Not sure how to change npm/pnpm versions in stackblitz but for me, signal handling works after upgrading npm and pnpm.

```sh
> npm -v && pnpm -v
10.2.4
9.5.0
```

**Comment by ryanxcharles** [comment]
> Hey Guys, any solution?

This issue affects me today with the latest version of pnpm. When I press ctrl+c inside a pnpm script, as described above, my script gets more than one SIGINT, and everything seems to get killed instantly, and my graceful shutdown code does not run.

I have decided to workaround this by using a shell script to run my application rather than a pnpm script.

Note that in my experience, both npm and pnpm have the same issue. The only workaround is to run node directly, or running node inside of a shell script.

**Comment by elclanrs** [comment]
We're experiencing the same problem. We run this script with `pnpm run ci` and when the build is killed with Ctrl+C, the SIGINT handler is killed prematurely and the cleanup does not happen:

```javascript
console.log('Running CI');
process.on('SIGINT', cleanup);
execSync('pnpm run build');
```

Handling the cleanup in the pnpm script itself works though, assuming that you exit with error code when cleanup is needed, but it's not very convenient:

```json
{
   "cleanup": "...",
   "ci": "node scripts/ci.js || cleanup"
}
```

**Comment by safareli** [comment]
I also experience this  with using pnpx

**Comment by simonduchen-volvo** [comment]
We are seeing this issue as well in our k8s environment when restarting pods and they are shutdown prematurely without our graceful shutdown handling. We have to start the apps with `node .` directly in the entrypoint for it to work.
Would be nice to have this work with pnpm though, since we now need separate entrypoint logic for different apps in our monorepo as some are Next.js apps and some a pure Node apps. Would be nice to just run `pnpm start` for all.

**Comment by alar0046** [comment]
This is still an issue with pnpm dlx

**Comment by vladshcherbin** [comment]
What a bizarre finding for such a common thing, mind-blowing 🤯

Workaround for users with modern nodejs versions is to run scripts directly bypassing pnpm:

```sh
node --run dev
```

**Comment by zkochan** [comment]
This is still reproducible on both pnpm v11 and v12, and fixes are open for both:

- pnpm v12 (Rust): https://github.com/pnpm/pnpm/pull/15106
- pnpm v11 (the relay lives in the `@pnpm/npm-lifecycle` package): https://github.com/pnpm/npm-lifecycle/pull/62, which then needs a release and a version bump in pnpm.

The cause is what was described above: `Ctrl+C` makes the terminal send `SIGINT` to the whole foreground process group, and pnpm then relayed the same `SIGINT` to the script, so the script got it twice. With the fix, pnpm does not relay the first `SIGINT` to a script in its own process group while pnpm is the foreground job of a terminal. A `SIGINT` sent to pnpm alone (`kill`, a container runtime, a service manager, none of which involve a terminal) is still relayed, so graceful shutdown in those environments keeps working. pnpm still waits for the script to finish shutting down before it exits.

---
Written by an agent (Claude Code, claude-fable-5-1).

**Comment by zkochan** [comment]
Update: the pnpm v11 fix (pnpm/npm-lifecycle#62) is merged upstream and reaches pnpm through https://github.com/pnpm/pnpm/pull/15113, which moves that package into this repository. The pnpm v12 fix remains https://github.com/pnpm/pnpm/pull/15106.

---
Written by an agent (Claude Code, claude-fable-5-1).
